// @tensorflow/tfjs-data/dist/iterators/microphone_iterator.js
"use strict";
/**
 * @license
 * Copyright 2019 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * =============================================================================
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var tfjs_core_1 = require("@tensorflow/tfjs-core");
var lazy_iterator_1 = require("./lazy_iterator");
/**
 * Provide a stream of tensors from microphone audio stream. The tensors are
 * representing audio data as frequency-domain spectrogram generated with
 * browser's native FFT. Tensors representing time-domain waveform is available
 * based on configuration. Only works in browser environment.
 */
var MicrophoneIterator = /** @class */ (function (_super) {
    __extends(MicrophoneIterator, _super);
    function MicrophoneIterator(microphoneConfig) {
        var _this = _super.call(this) || this;
        _this.microphoneConfig = microphoneConfig;
        _this.isClosed = false;
        _this.fftSize = microphoneConfig.fftSize || 1024;
        var fftSizeLog2 = Math.log2(_this.fftSize);
        if (_this.fftSize < 0 || fftSizeLog2 < 4 || fftSizeLog2 > 14 ||
            !Number.isInteger(fftSizeLog2)) {
            throw new Error("Invalid fftSize: it must be a power of 2 between " +
                ("2 to 4 and 2 to 14, but got " + _this.fftSize));
        }
        _this.numFrames = microphoneConfig.numFramesPerSpectrogram || 43;
        _this.sampleRateHz = microphoneConfig.sampleRateHz;
        _this.columnTruncateLength =
            microphoneConfig.columnTruncateLength || _this.fftSize;
        _this.audioTrackConstraints = microphoneConfig.audioTrackConstraints;
        _this.smoothingTimeConstant = microphoneConfig.smoothingTimeConstant || 0;
        _this.includeSpectrogram =
            microphoneConfig.includeSpectrogram === false ? false : true;
        _this.includeWaveform =
            microphoneConfig.includeWaveform === true ? true : false;
        if (!_this.includeSpectrogram && !_this.includeWaveform) {
            throw new Error('Both includeSpectrogram and includeWaveform are false. ' +
                'At least one type of data should be returned.');
        }
        return _this;
    }
    MicrophoneIterator.prototype.summary = function () {
        return "microphone";
    };
    // Construct a MicrophoneIterator and start the audio stream.
    MicrophoneIterator.create = function (microphoneConfig) {
        if (microphoneConfig === void 0) { microphoneConfig = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var microphoneIterator;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (tfjs_core_1.ENV.get('IS_NODE')) {
                            throw new Error('microphone API is only supported in browser environment.');
                        }
                        microphoneIterator = new MicrophoneIterator(microphoneConfig);
                        // Call async function start() to initialize the audio stream.
                        return [4 /*yield*/, microphoneIterator.start()];
                    case 1:
                        // Call async function start() to initialize the audio stream.
                        _a.sent();
                        return [2 /*return*/, microphoneIterator];
                }
            });
        });
    };
    // Start the audio stream and FFT.
    MicrophoneIterator.prototype.start = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, e_1, ctxConstructor, streamSource;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        _b.trys.push([0, 2, , 3]);
                        _a = this;
                        return [4 /*yield*/, navigator.mediaDevices.getUserMedia({
                                audio: this.audioTrackConstraints == null ? true :
                                    this.audioTrackConstraints,
                                video: false
                            })];
                    case 1:
                        _a.stream = _b.sent();
                        return [3 /*break*/, 3];
                    case 2:
                        e_1 = _b.sent();
                        throw new Error("Error thrown while initializing video stream: " + e_1.message);
                    case 3:
                        if (!this.stream) {
                            throw new Error('Could not obtain audio from microphone.');
                        }
                        ctxConstructor = 
                        // tslint:disable-next-line:no-any
                        window.AudioContext || window.webkitAudioContext;
                        this.audioContext = new ctxConstructor();
                        if (!this.sampleRateHz) {
                            // If sample rate is not provided, use the available sample rate on
                            // device.
                            this.sampleRateHz = this.audioContext.sampleRate;
                        }
                        else if (this.audioContext.sampleRate !== this.sampleRateHz) {
                            throw new Error("Mismatch in sampling rate: " +
                                ("Expected: " + this.sampleRateHz + "; ") +
                                ("Actual: " + this.audioContext.sampleRate));
                        }
                        streamSource = this.audioContext.createMediaStreamSource(this.stream);
                        this.analyser = this.audioContext.createAnalyser();
                        this.analyser.fftSize = this.fftSize * 2;
                        this.analyser.smoothingTimeConstant = this.smoothingTimeConstant;
                        streamSource.connect(this.analyser);
                        this.freqData = new Float32Array(this.fftSize);
                        this.timeData = new Float32Array(this.fftSize);
                        return [2 /*return*/];
                }
            });
        });
    };
    MicrophoneIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var spectrogramTensor, waveformTensor, audioDataQueue, freqData, timeData;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (this.isClosed) {
                            return [2 /*return*/, { value: null, done: true }];
                        }
                        return [4 /*yield*/, this.getAudioData()];
                    case 1:
                        audioDataQueue = _a.sent();
                        if (this.includeSpectrogram) {
                            freqData = this.flattenQueue(audioDataQueue.freqDataQueue);
                            spectrogramTensor = this.getTensorFromAudioDataArray(freqData, [this.numFrames, this.columnTruncateLength, 1]);
                        }
                        if (this.includeWaveform) {
                            timeData = this.flattenQueue(audioDataQueue.timeDataQueue);
                            waveformTensor = this.getTensorFromAudioDataArray(timeData, [this.numFrames * this.fftSize, 1]);
                        }
                        return [2 /*return*/, {
                                value: { 'spectrogram': spectrogramTensor, 'waveform': waveformTensor },
                                done: false
                            }];
                }
            });
        });
    };
    // Capture one result from the audio stream, and extract the value from
    // iterator.next() result.
    MicrophoneIterator.prototype.capture = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.next()];
                    case 1: return [2 /*return*/, (_a.sent()).value];
                }
            });
        });
    };
    MicrophoneIterator.prototype.getAudioData = function () {
        return __awaiter(this, void 0, void 0, function () {
            var freqDataQueue, timeDataQueue, currentFrames;
            var _this = this;
            return __generator(this, function (_a) {
                freqDataQueue = [];
                timeDataQueue = [];
                currentFrames = 0;
                return [2 /*return*/, new Promise(function (resolve) {
                        var intervalID = setInterval(function () {
                            if (_this.includeSpectrogram) {
                                _this.analyser.getFloatFrequencyData(_this.freqData);
                                // If the audio stream is initializing, return empty queue.
                                if (_this.freqData[0] === -Infinity) {
                                    resolve({ freqDataQueue: freqDataQueue, timeDataQueue: timeDataQueue });
                                }
                                freqDataQueue.push(_this.freqData.slice(0, _this.columnTruncateLength));
                            }
                            if (_this.includeWaveform) {
                                _this.analyser.getFloatTimeDomainData(_this.timeData);
                                timeDataQueue.push(_this.timeData.slice());
                            }
                            // Clean interval and return when all frames have been collected
                            if (++currentFrames === _this.numFrames) {
                                clearInterval(intervalID);
                                resolve({ freqDataQueue: freqDataQueue, timeDataQueue: timeDataQueue });
                            }
                        }, _this.fftSize / _this.sampleRateHz * 1e3);
                    })];
            });
        });
    };
    // Stop the audio stream and pause the iterator.
    MicrophoneIterator.prototype.stop = function () {
        this.isClosed = true;
        this.analyser.disconnect();
        this.audioContext.close();
        if (this.stream != null && this.stream.getTracks().length > 0) {
            this.stream.getTracks()[0].stop();
        }
    };
    // Override toArray() function to prevent collecting.
    MicrophoneIterator.prototype.toArray = function () {
        throw new Error('Can not convert infinite audio stream to array.');
    };
    // Return audio sampling rate in Hz
    MicrophoneIterator.prototype.getSampleRate = function () {
        return this.sampleRateHz;
    };
    MicrophoneIterator.prototype.flattenQueue = function (queue) {
        var frameSize = queue[0].length;
        var freqData = new Float32Array(queue.length * frameSize);
        queue.forEach(function (data, i) { return freqData.set(data, i * frameSize); });
        return freqData;
    };
    MicrophoneIterator.prototype.getTensorFromAudioDataArray = function (freqData, shape) {
        var vals = new Float32Array(tfjs_core_1.util.sizeFromShape(shape));
        // If the data is less than the output shape, the rest is padded with zeros.
        vals.set(freqData, vals.length - freqData.length);
        return tfjs_core_1.tensor(vals, shape);
    };
    return MicrophoneIterator;
}(lazy_iterator_1.LazyIterator));
exports.MicrophoneIterator = MicrophoneIterator;
