// @tensorflow/tfjs-data/dist/iterators/webcam_iterator.js
"use strict";
/**
 * @license
 * Copyright 2018 Google LLC. All Rights Reserved.
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * =============================================================================
 */
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __generator = (this && this.__generator) || function (thisArg, body) {
    var _ = { label: 0, sent: function() { if (t[0] & 1) throw t[1]; return t[1]; }, trys: [], ops: [] }, f, y, t, g;
    return g = { next: verb(0), "throw": verb(1), "return": verb(2) }, typeof Symbol === "function" && (g[Symbol.iterator] = function() { return this; }), g;
    function verb(n) { return function (v) { return step([n, v]); }; }
    function step(op) {
        if (f) throw new TypeError("Generator is already executing.");
        while (_) try {
            if (f = 1, y && (t = op[0] & 2 ? y["return"] : op[0] ? y["throw"] || ((t = y["return"]) && t.call(y), 0) : y.next) && !(t = t.call(y, op[1])).done) return t;
            if (y = 0, t) op = [op[0] & 2, t.value];
            switch (op[0]) {
                case 0: case 1: t = op; break;
                case 4: _.label++; return { value: op[1], done: false };
                case 5: _.label++; y = op[1]; op = [0]; continue;
                case 7: op = _.ops.pop(); _.trys.pop(); continue;
                default:
                    if (!(t = _.trys, t = t.length > 0 && t[t.length - 1]) && (op[0] === 6 || op[0] === 2)) { _ = 0; continue; }
                    if (op[0] === 3 && (!t || (op[1] > t[0] && op[1] < t[3]))) { _.label = op[1]; break; }
                    if (op[0] === 6 && _.label < t[1]) { _.label = t[1]; t = op; break; }
                    if (t && _.label < t[2]) { _.label = t[2]; _.ops.push(op); break; }
                    if (t[2]) _.ops.pop();
                    _.trys.pop(); continue;
            }
            op = body.call(thisArg, _);
        } catch (e) { op = [6, e]; y = 0; } finally { f = t = 0; }
        if (op[0] & 5) throw op[1]; return { value: op[0] ? op[1] : void 0, done: true };
    }
};
Object.defineProperty(exports, "__esModule", { value: true });
var tfjs_core_1 = require("@tensorflow/tfjs-core");
var lazy_iterator_1 = require("./lazy_iterator");
/**
 * Provide a stream of image tensors from webcam video stream. Only works in
 * browser environment.
 */
var WebcamIterator = /** @class */ (function (_super) {
    __extends(WebcamIterator, _super);
    function WebcamIterator(webcamVideoElement, webcamConfig) {
        var _this = _super.call(this) || this;
        _this.webcamVideoElement = webcamVideoElement;
        _this.webcamConfig = webcamConfig;
        _this.isClosed = true;
        _this.resize = false;
        if (_this.needToResize()) {
            _this.resize = true;
            _this.cropSize =
                [_this.webcamConfig.resizeHeight, _this.webcamConfig.resizeWidth];
            _this.cropBoxInd = tfjs_core_1.tensor1d([0], 'int32');
            if (_this.webcamConfig.centerCrop) {
                // Calculate the box based on resizing shape.
                var widthCroppingRatio = _this.webcamConfig.resizeWidth * 1.0 / _this.webcamVideoElement.width;
                var heightCroppingRatio = _this.webcamConfig.resizeHeight * 1.0 /
                    _this.webcamVideoElement.height;
                var widthCropStart = (1 - widthCroppingRatio) / 2;
                var heightCropStart = (1 - heightCroppingRatio) / 2;
                var widthCropEnd = widthCropStart + widthCroppingRatio;
                var heightCropEnd = heightCroppingRatio + heightCropStart;
                _this.cropBox = tfjs_core_1.tensor2d([heightCropStart, widthCropStart, heightCropEnd, widthCropEnd], [1, 4]);
            }
            else {
                _this.cropBox = tfjs_core_1.tensor2d([0, 0, 1, 1], [1, 4]);
            }
        }
        return _this;
    }
    WebcamIterator.prototype.summary = function () {
        return "webcam";
    };
    // Construct a WebcamIterator and start it's video stream.
    WebcamIterator.create = function (webcamVideoElement, webcamConfig) {
        if (webcamConfig === void 0) { webcamConfig = {}; }
        return __awaiter(this, void 0, void 0, function () {
            var webcamIterator;
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0:
                        if (tfjs_core_1.ENV.get('IS_NODE')) {
                            throw new Error('tf.data.webcam is only supported in browser environment.');
                        }
                        if (!webcamVideoElement) {
                            // If webcam video element is not provided, create a hidden video element
                            // with provided width and height.
                            webcamVideoElement = document.createElement('video');
                            if (!webcamConfig.resizeWidth || !webcamConfig.resizeHeight) {
                                throw new Error('Please provide webcam video element, or resizeWidth and ' +
                                    'resizeHeight to create a hidden video element.');
                            }
                            webcamVideoElement.width = webcamConfig.resizeWidth;
                            webcamVideoElement.height = webcamConfig.resizeHeight;
                        }
                        webcamIterator = new WebcamIterator(webcamVideoElement, webcamConfig);
                        // Call async function to initialize the video stream.
                        return [4 /*yield*/, webcamIterator.start()];
                    case 1:
                        // Call async function to initialize the video stream.
                        _a.sent();
                        return [2 /*return*/, webcamIterator];
                }
            });
        });
    };
    // Async function to start video stream.
    WebcamIterator.prototype.start = function () {
        return __awaiter(this, void 0, void 0, function () {
            var _a, e_1;
            var _this = this;
            return __generator(this, function (_b) {
                switch (_b.label) {
                    case 0:
                        if (this.webcamConfig.facingMode) {
                            tfjs_core_1.util.assert((this.webcamConfig.facingMode === 'user') ||
                                (this.webcamConfig.facingMode === 'environment'), function () {
                                return "Invalid webcam facing mode: " + _this.webcamConfig.facingMode + ". " +
                                    "Please provide 'user' or 'environment'";
                            });
                        }
                        _b.label = 1;
                    case 1:
                        _b.trys.push([1, 3, , 4]);
                        _a = this;
                        return [4 /*yield*/, navigator.mediaDevices.getUserMedia({
                                video: {
                                    deviceId: this.webcamConfig.deviceId,
                                    facingMode: this.webcamConfig.facingMode ?
                                        this.webcamConfig.facingMode :
                                        'user',
                                    width: this.webcamVideoElement.width,
                                    height: this.webcamVideoElement.height
                                }
                            })];
                    case 2:
                        _a.stream = _b.sent();
                        return [3 /*break*/, 4];
                    case 3:
                        e_1 = _b.sent();
                        // Modify the error message but leave the stack trace intact
                        e_1.message = "Error thrown while initializing video stream: " + e_1.message;
                        throw e_1;
                    case 4:
                        if (!this.stream) {
                            throw new Error('Could not obtain video from webcam.');
                        }
                        // Older browsers may not have srcObject
                        try {
                            this.webcamVideoElement.srcObject = this.stream;
                        }
                        catch (error) {
                            console.log(error);
                            this.webcamVideoElement.src = window.URL.createObjectURL(this.stream);
                        }
                        // Start the webcam video stream
                        this.webcamVideoElement.play();
                        this.isClosed = false;
                        return [2 /*return*/, new Promise(function (resolve) {
                                // Add event listener to make sure the webcam has been fully initialized.
                                _this.webcamVideoElement.onloadedmetadata = function () {
                                    resolve();
                                };
                            })];
                }
            });
        });
    };
    WebcamIterator.prototype.next = function () {
        return __awaiter(this, void 0, void 0, function () {
            var img;
            return __generator(this, function (_a) {
                if (this.isClosed) {
                    return [2 /*return*/, { value: null, done: true }];
                }
                try {
                    img = tfjs_core_1.browser.fromPixels(this.webcamVideoElement);
                }
                catch (e) {
                    throw new Error("Error thrown converting video to pixels: " + JSON.stringify(e));
                }
                if (this.resize) {
                    try {
                        return [2 /*return*/, { value: this.cropAndResizeFrame(img), done: false }];
                    }
                    catch (e) {
                        throw new Error("Error thrown cropping the video: " + e.message);
                    }
                    finally {
                        img.dispose();
                    }
                }
                else {
                    return [2 /*return*/, { value: img, done: false }];
                }
                return [2 /*return*/];
            });
        });
    };
    WebcamIterator.prototype.needToResize = function () {
        // If resizeWidth and resizeHeight are provided, and different from the
        // width and height of original HTMLVideoElement, then resizing and cropping
        // is required.
        if (this.webcamConfig.resizeWidth && this.webcamConfig.resizeHeight &&
            (this.webcamVideoElement.width !== this.webcamConfig.resizeWidth ||
                this.webcamVideoElement.height !== this.webcamConfig.resizeHeight)) {
            return true;
        }
        return false;
    };
    // Cropping and resizing each frame based on config
    WebcamIterator.prototype.cropAndResizeFrame = function (img) {
        var _this = this;
        return tfjs_core_1.tidy(function () {
            var expandedImage = img.toFloat().expandDims(0);
            var resizedImage;
            resizedImage = tfjs_core_1.image.cropAndResize(expandedImage, _this.cropBox, _this.cropBoxInd, _this.cropSize, 'bilinear');
            // Extract image from batch cropping.
            var shape = resizedImage.shape;
            return resizedImage.reshape(shape.slice(1));
        });
    };
    // Capture one frame from the video stream, and extract the value from
    // iterator.next() result.
    WebcamIterator.prototype.capture = function () {
        return __awaiter(this, void 0, void 0, function () {
            return __generator(this, function (_a) {
                switch (_a.label) {
                    case 0: return [4 /*yield*/, this.next()];
                    case 1: return [2 /*return*/, (_a.sent()).value];
                }
            });
        });
    };
    // Stop the video stream and pause webcam iterator.
    WebcamIterator.prototype.stop = function () {
        var tracks = this.stream.getTracks();
        tracks.forEach(function (track) { return track.stop(); });
        try {
            this.webcamVideoElement.srcObject = null;
        }
        catch (error) {
            console.log(error);
            this.webcamVideoElement.src = null;
        }
        this.isClosed = true;
    };
    // Override toArray() function to prevent collecting.
    WebcamIterator.prototype.toArray = function () {
        throw new Error('Can not convert infinite video stream to array.');
    };
    return WebcamIterator;
}(lazy_iterator_1.LazyIterator));
exports.WebcamIterator = WebcamIterator;
